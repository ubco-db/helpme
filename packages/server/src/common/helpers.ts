import { Request } from 'express';

export function getCookie(req: Request, cookieName: string): string {
  const cookieHeader = req.headers.cookie;
  if (!cookieHeader) {
    return null;
  }

  const cookies: string[] = cookieHeader.split(';');
  for (const cookie of cookies) {
    const [name, value] = cookie.split('=');
    if (name.trim() === cookieName) {
      return value;
    }
  }
}
