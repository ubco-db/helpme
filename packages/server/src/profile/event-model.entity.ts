import { Exclude } from 'class-transformer';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { UserModel } from './user.entity';
import { QueueModel } from '../queue/queue.entity';

/**
 * Represents an Event in the EventModel table, used for advanced metrics.
 */
export enum EventType {
  TA_CHECKED_IN = 'taCheckedIn',
  TA_CHECKED_OUT = 'taCheckedOut',
  TA_CHECKED_OUT_FORCED = 'taCheckedOutForced',
  TA_CHECKED_OUT_EVENT_END = 'taCheckedOutEventEnd',
}

@Entity('event_model')
export class EventModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  time: Date;

  @Column({ type: 'enum', enum: EventType })
  eventType: EventType;

  @ManyToOne((type) => UserModel, (user) => user.events, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'userId' })
  user: UserModel;

  @Column({ nullable: true })
  @Exclude()
  userId: number;

  @ManyToOne((type) => CourseModel, (course) => course.events, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  @Exclude()
  courseId: number;

  @Column({ nullable: true })
  @Exclude()
  queueId: number;

  @ManyToOne((type) => QueueModel, { onDelete: 'SET NULL' })
  @JoinColumn({ name: 'queueId' })
  queue: QueueModel;
}
