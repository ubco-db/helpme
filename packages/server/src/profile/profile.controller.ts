import {
  DesktopNotifPartial,
  ERROR_MESSAGES,
  GetProfileResponse,
  UpdateProfileParams,
  AccountType,
} from '@koh/common';
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
  Res,
  ServiceUnavailableException,
  UploadedFile,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import * as checkDiskSpace from 'check-disk-space';
import { Response } from 'express';
import * as fs from 'fs';
import { pick } from 'lodash';
import { memoryStorage } from 'multer';
import * as path from 'path';
import * as sharp from 'sharp';
import { JwtAuthGuard } from '../guards/jwt-auth.guard';
import { User } from '../decorators/user.decorator';
import { UserModel } from './user.entity';
import { OrganizationService } from '../organization/organization.service';
import { EmailVerifiedGuard } from 'guards/email-verified.guard';
import { minutes, SkipThrottle, Throttle } from '@nestjs/throttler';

@Controller('profile')
export class ProfileController {
  constructor(private organizationService: OrganizationService) {}

  // Don't throttle this endpoint since the middleware calls this for every page (and if it prefetches like 30 pages, it will hit the throttle limit and can cause issue for the user)
  @SkipThrottle()
  @Get()
  @UseGuards(JwtAuthGuard)
  async get(
    @User(['courses', 'courses.course', 'desktopNotifs', 'chat_token'])
    user: UserModel,
  ): Promise<GetProfileResponse> {
    if (user === null || user === undefined) {
      console.error(ERROR_MESSAGES.profileController.accountNotAvailable);
      throw new HttpException(
        ERROR_MESSAGES.profileController.accountNotAvailable,
        HttpStatus.NOT_FOUND,
      );
    }

    if (user.accountDeactivated) {
      throw new HttpException(
        ERROR_MESSAGES.profileController.accountDeactivated,
        HttpStatus.FORBIDDEN,
      );
    }

    const courses = user.courses
      ? user.courses
          .filter((userCourse) => userCourse?.course?.enabled)
          .map((userCourse) => {
            return {
              course: {
                id: userCourse.courseId,
                name: userCourse.course.name,
              },
              role: userCourse.role,
            };
          })
      : [];

    const desktopNotifs: DesktopNotifPartial[] = user.desktopNotifs
      ? user.desktopNotifs.map((d) => ({
          endpoint: d.endpoint,
          id: d.id,
          createdAt: d.createdAt,
          name: d.name,
        }))
      : [];

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'sid',
      'firstName',
      'lastName',
      'photoURL',
      'defaultMessage',
      'includeDefaultMessage',
      'desktopNotifsEnabled',
      'insights',
      'userRole',
      'accountType',
      'emailVerified',
      'chat_token',
      'readChangeLog',
    ]);

    if (userResponse === null || userResponse === undefined) {
      console.error(ERROR_MESSAGES.profileController.userResponseNotFound);
      throw new HttpException(
        ERROR_MESSAGES.profileController.userResponseNotFound,
        HttpStatus.NOT_FOUND,
      );
    }

    const userOrganization =
      await this.organizationService.getOrganizationAndRoleByUserId(user.id);

    const organization = pick(userOrganization, [
      'id',
      'orgId',
      'organizationName',
      'organizationDescription',
      'organizationLogoUrl',
      'organizationBannerUrl',
      'organizationRole',
    ]);

    return {
      ...userResponse,
      courses,
      desktopNotifs,
      organization,
    };
  }

  @Patch()
  @UseGuards(JwtAuthGuard, EmailVerifiedGuard)
  async patch(
    @Res() res: Response,
    @Body() userPatch: UpdateProfileParams,
    @User()
    user: UserModel,
  ): Promise<Response<GetProfileResponse>> {
    if (user.accountType !== AccountType.LEGACY && userPatch.email) {
      return res
        .status(HttpStatus.BAD_REQUEST)
        .send({ message: ERROR_MESSAGES.profileController.cannotUpdateEmail });
    }

    if (userPatch.email && userPatch.email !== user.email) {
      const email = await UserModel.findOne({
        where: {
          email: userPatch.email,
        },
      });

      if (email) {
        return res
          .status(HttpStatus.BAD_REQUEST)
          .send({ message: ERROR_MESSAGES.profileController.emailAlreadyInDb });
      }
    }

    if (userPatch.sid && userPatch.sid !== user.sid) {
      const sid = await UserModel.findOne({
        where: {
          sid: userPatch.sid,
        },
      });

      if (sid) {
        return res
          .status(HttpStatus.BAD_REQUEST)
          .send({ message: ERROR_MESSAGES.profileController.sidAlreadyInDb });
      }
    }

    user = Object.assign(user, userPatch);

    await user
      .save()
      .then(() => {
        return this.get(user);
      })
      .catch((e) => {
        console.log(e);
      });

    return res.status(200).send({ message: 'Profile updated successfully' });
  }

  // Only 10 calls allowed in 1 minute
  @Throttle({ default: { limit: 10, ttl: minutes(1) } })
  @Post('/upload_picture')
  @UseGuards(JwtAuthGuard, EmailVerifiedGuard)
  @UseInterceptors(
    FileInterceptor('file', {
      storage: memoryStorage(),
    }),
  )
  async uploadImage(
    @UploadedFile() file: Express.Multer.File,
    @User() user: UserModel,
    @Res() response: Response,
  ): Promise<void> {
    try {
      /*
       * The second check is for google accounts, which have a photoURL that is a external link
       */
      if (user.photoURL && !user.photoURL.startsWith('http')) {
        try {
          fs.unlinkSync(path.join(process.env.UPLOAD_LOCATION, user.photoURL));
        } catch (e) {
          console.error(
            'Error deleting previous picture at : ' +
              user.photoURL +
              '\n Perhaps the previous image was deleted or the database is out of sync with the uploads directory for some reason.' +
              '\n Will remove this entry from the database and continue.',
          );
        }
      }

      const spaceLeft = await checkDiskSpace(path.parse(process.cwd()).root);

      if (spaceLeft.free < 1_000_000_000) {
        // if less than a gigabyte left
        throw new ServiceUnavailableException(
          ERROR_MESSAGES.profileController.noDiskSpace,
        );
      }
      const fileName = user.id + '-' + Date.now().toString() + '.webp';

      // Create the upload location if it doesn't exist
      if (!fs.existsSync(process.env.UPLOAD_LOCATION)) {
        fs.mkdirSync(process.env.UPLOAD_LOCATION, { recursive: true });
      }

      const targetPath = path.join(process.env.UPLOAD_LOCATION, fileName);

      try {
        await sharp(file.buffer).resize(256).webp().toFile(targetPath);
        user.photoURL = fileName;
      } catch (err) {
        console.error('Error processing image:', err);
      }
      await user.save();
      response
        .status(200)
        .send({ message: 'Image uploaded successfully', fileName });
    } catch (error) {
      response
        .status(500)
        .send({ message: 'Image upload failed', error: error.message });
    }
  }

  @Get('/get_picture/:photoURL')
  @UseGuards(JwtAuthGuard, EmailVerifiedGuard)
  async getImage(
    @Param('photoURL') photoURL: string,
    @Res() res: Response,
  ): Promise<void> {
    fs.stat(
      path.join(process.env.UPLOAD_LOCATION, photoURL),
      async (err, stats) => {
        if (err) {
          return res
            .status(HttpStatus.NOT_FOUND)
            .send({ message: 'File not found' });
        }
        if (stats) {
          res.set('Content-Type', 'image/webp');
          res.sendFile(
            photoURL,
            { root: process.env.UPLOAD_LOCATION },
            (sendFileError) => {
              if (sendFileError) {
                return res
                  .status(HttpStatus.INTERNAL_SERVER_ERROR)
                  .send({ message: 'Error serving the image.' });
              }
            },
          );
        }
      },
    );
  }

  @Delete('/delete_profile_picture')
  @UseGuards(JwtAuthGuard, EmailVerifiedGuard)
  async deleteProfilePicture(
    @User() user: UserModel,
    @Res() res: Response,
  ): Promise<Response> {
    if (!user?.photoURL) {
      return res
        .status(HttpStatus.NOT_FOUND)
        .send({ message: 'No profile picture to delete' });
    }

    if (user.photoURL.startsWith('http')) {
      user.photoURL = null;
      await user.save();
      return res
        .status(HttpStatus.OK)
        .send({ message: 'Profile picture deleted successfully' });
    } else {
      fs.unlink(
        process.env.UPLOAD_LOCATION + '/' + user.photoURL,
        async (err) => {
          if (err) {
            const errMessage =
              'Error deleting previous picture at : ' +
              user.photoURL +
              'the previous image was at an invalid location?';
            console.error(errMessage, err);
            throw new BadRequestException(errMessage);
          } else {
            user.photoURL = null;
            await user.save();
            return res
              .status(HttpStatus.OK)
              .send({ message: 'Profile picture deleted successfully' });
          }
        },
      );
    }
  }

  @Patch('/read_changelog')
  @UseGuards(JwtAuthGuard, EmailVerifiedGuard)
  async readChangelogs(
    @User() user: UserModel,
    @Res() res: Response,
  ): Promise<Response> {
    try {
      user.readChangeLog = true;
      await user.save();
      return res
        .status(HttpStatus.OK)
        .send({ message: 'Changelogs read successfully' });
    } catch (error) {
      return res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .send({ message: 'Error reading changelogs' });
    }
  }
}
