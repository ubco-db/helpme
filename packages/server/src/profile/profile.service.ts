import {
  AccountType,
  DesktopNotifPartial,
  ERROR_MESSAGES,
  GetProfileResponse,
  Role,
  UpdateProfileParams,
  User,
} from '@koh/common';
import {
  BadRequestException,
  HttpException,
  HttpStatus,
  Injectable,
  NotFoundException,
  ServiceUnavailableException,
} from '@nestjs/common';
import { UserModel } from './user.entity';
import { pick } from 'lodash';
import { OrganizationService } from '../organization/organization.service';
import * as checkDiskSpaceModule from 'check-disk-space';
import * as path from 'path';
import * as fs from 'fs';
import * as sharpModule from 'sharp';

const checkDiskSpace =
  (checkDiskSpaceModule as any).default || checkDiskSpaceModule;
const sharp = (sharpModule as any).default || sharpModule;

@Injectable()
export class ProfileService {
  constructor(private organizationService: OrganizationService) {}

  async getProfile(user: UserModel): Promise<User> {
    const courses = user.courses
      ? user.courses
          .filter(
            (userCourse) =>
              userCourse?.course?.enabled ||
              userCourse?.role === Role.PROFESSOR,
          )
          .map((userCourse) => {
            return {
              course: {
                id: userCourse.courseId,
                name: userCourse.course.name,
                semesterId: userCourse.course.semesterId,
                enabled: userCourse.course.enabled,
                sectionGroupName: userCourse.course.sectionGroupName,
              },
              role: userCourse.role,
              favourited: userCourse.favourited,
            };
          })
      : [];

    const desktopNotifs: DesktopNotifPartial[] = user.desktopNotifs
      ? user.desktopNotifs.map((d) => ({
          endpoint: d.endpoint,
          id: d.id,
          createdAt: d.createdAt,
          name: d.name,
        }))
      : [];

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'sid',
      'firstName',
      'lastName',
      'photoURL',
      'defaultMessage',
      'includeDefaultMessage',
      'desktopNotifsEnabled',
      'insights',
      'userRole',
      'accountType',
      'emailVerified',
      'chat_token',
      'readChangeLog',
    ]);

    if (userResponse === null || userResponse === undefined) {
      console.error(ERROR_MESSAGES.profileController.userResponseNotFound);
      throw new HttpException(
        ERROR_MESSAGES.profileController.userResponseNotFound,
        HttpStatus.NOT_FOUND,
      );
    }

    const userOrganization =
      await this.organizationService.getOrganizationAndRoleByUserId(user.id);

    const organization = pick(userOrganization, [
      'id',
      'orgId',
      'organizationName',
      'organizationDescription',
      'organizationLogoUrl',
      'organizationBannerUrl',
      'organizationRole',
    ]);

    const profile = {
      ...userResponse,
      courses,
      desktopNotifs,
      organization,
    } as User;

    return profile;
  }

  async uploadUserProfileImage(
    file: Express.Multer.File,
    user: UserModel,
  ): Promise<string> {
    try {
      // Remove previous profile picture (if it's not an external URL)
      if (user.photoURL && !user.photoURL.startsWith('http')) {
        this.deletePreviousImage(user.photoURL);
      }

      // Check disk space before proceeding
      const spaceLeft = await checkDiskSpace(path.parse(process.cwd()).root);
      if (spaceLeft.free < 1_000_000_000) {
        throw new ServiceUnavailableException(
          ERROR_MESSAGES.profileController.noDiskSpace,
        );
      }

      const fileName = `${user.id}-${Date.now()}.webp`;
      const targetPath = path.join(process.env.UPLOAD_LOCATION, fileName);

      // Ensure the upload directory exists
      if (!fs.existsSync(process.env.UPLOAD_LOCATION)) {
        fs.mkdirSync(process.env.UPLOAD_LOCATION, { recursive: true });
      }

      // Process and save the image
      await sharp(file.buffer).resize(256).webp().toFile(targetPath);
      user.photoURL = fileName;
      await user.save();

      return fileName;
    } catch (error) {
      console.error('Error processing image:', error);
      throw error;
    }
  }

  private deletePreviousImage(filePath: string): void {
    try {
      fs.unlinkSync(path.join(process.env.UPLOAD_LOCATION, filePath));
    } catch (e) {
      console.error(
        `Error deleting previous picture at: ${filePath}\n` +
          `Perhaps the file was already deleted or the database is out of sync with the uploads directory.\n` +
          `Will remove this entry from the database and continue.`,
      );
    }
  }

  async updateUserProfile(
    user: UserModel,
    userPatch: UpdateProfileParams,
  ): Promise<GetProfileResponse> {
    if (user.accountType !== AccountType.LEGACY && userPatch.email) {
      throw new BadRequestException(
        ERROR_MESSAGES.profileController.cannotUpdateEmail,
      );
    }

    // Check if the new email is already in use
    if (userPatch.email && userPatch.email !== user.email) {
      const emailExists = await UserModel.findOne({
        where: { email: userPatch.email },
      });
      if (emailExists) {
        throw new BadRequestException(
          ERROR_MESSAGES.profileController.emailAlreadyInDb,
        );
      }
    }

    // Check if the new student ID (sid) is already in use
    if (userPatch.sid && userPatch.sid !== user.sid) {
      const sidExists = await UserModel.findOne({
        where: { sid: userPatch.sid },
      });
      if (sidExists) {
        throw new BadRequestException(
          ERROR_MESSAGES.profileController.sidAlreadyInDb,
        );
      }
    }

    // Update user with new data
    Object.assign(user, userPatch);

    // Save updated user
    await user.save();

    return user;
  }

  async removeProfilePicture(user: UserModel): Promise<void> {
    if (!user?.photoURL) {
      throw new NotFoundException('No profile picture to delete');
    }

    if (user.photoURL.startsWith('http')) {
      user.photoURL = null;
      await user.save();
      return;
    }

    const filePath = path.join(process.env.UPLOAD_LOCATION, user.photoURL);

    try {
      await fs.promises.unlink(filePath);
      user.photoURL = null;
      await user.save();
    } catch (err) {
      console.error(`Error deleting profile picture at: ${filePath}`, err);
      throw new BadRequestException(
        'Error deleting profile picture. The file may not exist.',
      );
    }
  }
}
