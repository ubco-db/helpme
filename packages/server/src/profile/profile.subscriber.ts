import {
  DataSource,
  EntitySubscriberInterface,
  EventSubscriber,
  RemoveEvent,
  UpdateEvent,
} from 'typeorm';
import { UserModel } from './user.entity';
import { RedisProfileService } from 'redisProfile/redis-profile.service';

@EventSubscriber()
export class ProfileSubscriber implements EntitySubscriberInterface<UserModel> {
  constructor(
    private readonly redisProfileService: RedisProfileService,
    dataSource: DataSource,
  ) {
    dataSource.subscribers.push(this);
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listenTo() {
    return UserModel;
  }

  async afterUpdate(event: UpdateEvent<UserModel>): Promise<void> {
    // delete cached profile whenever user is updated
    await this.redisProfileService.deleteProfile(`u:${event.entity.id}`);
  }

  async beforeRemove(event: RemoveEvent<UserModel>): Promise<void> {
    // due to cascades entity is not guaranteed to be loaded
    if (event.entity) {
      await this.redisProfileService.deleteProfile(`u:${event.entity.id}`);
    }
  }
}
