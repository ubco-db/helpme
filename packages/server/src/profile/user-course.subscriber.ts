import {
  DataSource,
  EntitySubscriberInterface,
  EventSubscriber,
  InsertEvent,
  RemoveEvent,
  UpdateEvent,
} from 'typeorm';
import { UserCourseModel } from './user-course.entity';
import { RedisProfileService } from 'redisProfile/redis-profile.service';

@EventSubscriber()
export class UserCourseSubscriber
  implements EntitySubscriberInterface<UserCourseModel>
{
  constructor(
    private readonly redisProfileService: RedisProfileService,
    dataSource: DataSource,
  ) {
    dataSource.subscribers.push(this);
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listenTo() {
    return UserCourseModel;
  }

  async afterUpdate(event: UpdateEvent<UserCourseModel>): Promise<void> {
    // delete cached profile whenever user course is updated
    let userId: number;

    if (event.entity.user?.id) {
      userId = event.entity.user.id;
    } else if (event.entity.userId) {
      userId = event.entity.userId;
    } else {
      // If user relationship is not loaded, we need to query for it
      const userCourse = await UserCourseModel.findOne({
        where: { id: event.entity.id },
        relations: ['user'],
      });
      if (userCourse?.user?.id) {
        userId = userCourse.user.id;
      }
    }

    if (userId) {
      await this.redisProfileService.deleteProfile(`u:${userId}`);
    }
  }

  async beforeRemove(event: RemoveEvent<UserCourseModel>): Promise<void> {
    // due to cascades entity is not guaranteed to be loaded
    if (!event.entity) {
      return;
    }

    let userId: number;

    if (event.entity.user?.id) {
      userId = event.entity.user.id;
    } else if (event.entity.userId) {
      userId = event.entity.userId;
    } else {
      // If user relationship is not loaded, we need to query for it
      const userCourse = await UserCourseModel.findOne({
        where: { id: event.entity.id },
        relations: ['user'],
      });
      if (userCourse?.user?.id) {
        userId = userCourse.user.id;
      }
    }

    if (userId) {
      await this.redisProfileService.deleteProfile(`u:${userId}`);
    }
  }

  async afterInsert(event: InsertEvent<UserCourseModel>): Promise<void> {
    let userId: number;

    if (event.entity.user?.id) {
      userId = event.entity.user.id;
    } else if (event.entity.userId) {
      userId = event.entity.userId;
    } else {
      // If user relationship is not loaded, we need to query for it
      const userCourse = await UserCourseModel.findOne({
        where: { id: event.entity.id },
        relations: ['user'],
      });
      if (userCourse?.user?.id) {
        userId = userCourse.user.id;
      }
    }

    if (userId) {
      await this.redisProfileService.deleteProfile(`u:${userId}`);
    }
  }
}
