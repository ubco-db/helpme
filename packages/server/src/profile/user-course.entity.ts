import { Role } from '@koh/common';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { UserModel } from './user.entity';
import { InsightDashboardModel } from '../insights/dashboard.entity';

@Entity('user_course_model')
export class UserCourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => UserModel, (user) => user.courses)
  @JoinColumn({ name: 'userId' })
  user: UserModel;

  @Column({ nullable: true })
  userId: number;

  @ManyToOne((type) => CourseModel, (course) => course.userCourses)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  courseId: number;

  @Column({ type: 'enum', enum: Role, default: Role.STUDENT })
  role: Role;

  // If this expires
  @Column({ nullable: true, default: false })
  expires: boolean;

  @OneToMany(
    (type) => InsightDashboardModel,
    (insightDashboard) => insightDashboard.userCourse,
  )
  insightDashboard?: InsightDashboardModel[];

  /* This represents the notes that professors can put on their TA's profiles*/
  @Column({ type: 'text', nullable: true })
  TANotes: string;

  @Column({ nullable: true, default: true })
  favourited: boolean;
}
