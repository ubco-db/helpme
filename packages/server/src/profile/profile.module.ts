import { Module } from '@nestjs/common';
import { ProfileController } from './profile.controller';
import { NotificationModule } from '../notification/notification.module';
import { LoginModule } from 'login/login.module';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { JwtStrategy } from './jwt.strategy';
import { MailModule } from 'mail/mail.module';
import { OrganizationModule } from '../organization/organization.module';
import { RedisProfileService } from '../redisProfile/redis-profile.service';
import { RedisProfileModule } from '../redisProfile/redis-profile.module';
import { ProfileService } from './profile.service';

@Module({
  imports: [
    NotificationModule,
    LoginModule,
    MailModule,
    RedisProfileModule,
    JwtModule.registerAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get('JWT_SECRET'),
      }),
    }),
    OrganizationModule,
  ],
  controllers: [ProfileController],
  providers: [JwtStrategy, ProfileService, RedisProfileService],
  exports: [ProfileService, RedisProfileService],
})
export class ProfileModule {}
