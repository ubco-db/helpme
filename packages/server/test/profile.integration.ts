import {
  StudentCourseFactory,
  UserFactory,
  CourseFactory,
  OrganizationFactory,
  ChatTokenFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { ProfileModule } from '../src/profile/profile.module';
import { DesktopNotifModel } from 'notification/desktop-notif.entity';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import { AccountType } from '@koh/common';

describe('Profile Integration', () => {
  const { supertest } = setupIntegrationTest(ProfileModule);

  describe('GET /profile', () => {
    it('returns the logged-in user profile', async () => {
      const organization = await OrganizationFactory.create();
      const user = await UserFactory.create();
      const fundies = await CourseFactory.create({ name: 'CS 2500' });
      await StudentCourseFactory.create({ course: fundies, user });
      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      await ChatTokenFactory.create({ user, token: 'test' });

      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);
      expect(res.body).toMatchSnapshot();
    });

    it('returns account deactivated error', async () => {
      const user = await UserFactory.create({ accountDeactivated: true });
      await supertest({ userId: user.id }).get('/profile').expect(403);
    });

    it('returns only userCourses where course is enabled', async () => {
      const user = await UserFactory.create();
      const fundies = await CourseFactory.create({ name: 'CS 2500' });
      const nonEnabledCourse = await CourseFactory.create({
        name: 'CS 4900',
        enabled: false,
      });
      await StudentCourseFactory.create({ course: fundies, user });
      await StudentCourseFactory.create({ course: nonEnabledCourse, user });

      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);

      expect(res.body.courses).toEqual([
        {
          course: {
            id: 1,
            name: 'CS 2500',
          },
          role: 'student',
        },
      ]);
    });

    it('returns desktop notif information', async () => {
      const user = await UserFactory.create();
      const dn = await DesktopNotifModel.create({
        user,
        auth: '',
        p256dh: '',
        endpoint: 'abc',
        name: 'firefox',
      }).save();
      await dn.reload();
      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);
      expect(res.body.desktopNotifs).toEqual([
        {
          createdAt: expect.any(String),
          name: 'firefox',
          id: dn.id,
          endpoint: dn.endpoint,
        },
      ]);
    });

    it('returns 401 when not logged in', async () => {
      await UserFactory.create();
      await supertest().get('/profile').expect(401);
    });
  });

  describe('DELETE /profile/delete_profile_picture', () => {
    it("should delete profile picture from database only if it's URL", async () => {
      const user = await UserFactory.create({
        photoURL: 'https://www.google.com',
      });
      await supertest({ userId: user.id })
        .delete('/profile/delete_profile_picture')
        .expect(200);
    });
  });

  describe('PATCH /profile', () => {
    it('enables desktop notifs', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
      });
      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ desktopNotifsEnabled: true })
        .expect(200);
      expect(res.body).toMatchObject({
        message: 'Profile updated successfully',
      });
    });

    it('should not update profile settings if email is included and account is not legacy', async () => {
      const user = await UserFactory.create({
        email: 'test@ubc.ca',
        accountType: AccountType.GOOGLE,
      });

      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: 'test@ubc.ca', firstName: 'test' });

      expect(res.status).toEqual(400);
      expect(res.body).toEqual({
        message: 'Email cannot be updated',
      });
    });

    it('lets ta change default teams message', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.defaultMessage).toEqual(null);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ defaultMessage: "Hello! It's me :D" })
        .expect(200);
      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.defaultMessage).toEqual("Hello! It's me :D");
    });

    it('lets ta change includeDefaultMessage', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.includeDefaultMessage).toEqual(true);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ includeDefaultMessage: false })
        .expect(200);
      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.includeDefaultMessage).toEqual(false);
    });

    it('lets user change email', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      const newEmail = 'new_test_email@ubc.ca';

      expect(profile.body?.email).toEqual(user.email);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: newEmail })
        .expect(200);

      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.email).toEqual(newEmail);
    });

    it('fails to change user email when email is used by another user', async () => {
      const user = await UserFactory.create();
      await UserFactory.create({ email: 'test@ubc.ca' });

      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: 'test@ubc.ca' })
        .expect(400);
    });
  });
});
