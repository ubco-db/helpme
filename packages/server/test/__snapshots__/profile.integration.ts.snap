// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Integration GET /profile returns the logged-in user profile 1`] = `
{
  "accountType": "legacy",
  "chat_token": {
    "id": 1,
    "max_uses": 30,
    "token": "test",
    "used": 0,
  },
  "courses": [
    {
      "course": {
        "id": 1,
        "name": "CS 2500",
      },
      "role": "student",
    },
  ],
  "defaultMessage": null,
  "desktopNotifs": [],
  "desktopNotifsEnabled": false,
  "email": "user@ubc.ca",
  "emailVerified": true,
  "firstName": "User",
  "id": 1,
  "includeDefaultMessage": true,
  "insights": [
    "TotalStudents",
    "TotalQuestionsAsked",
    "MedianWaitTime",
    "QuestionTypeBreakdown",
    "MostActiveStudents",
    "QuestionToStudentRatio",
    "MedianHelpingTime",
    "AverageTimesByWeekDay",
    "HumanVsChatbot",
    "HumanVsChatbotVotes",
    "MostActiveTimes",
    "HelpSeekingOverTime",
    "StaffEfficiency",
    "StaffTotalHelped",
    "StaffQuestionTimesByDay",
    "StaffWorkload",
  ],
  "lastName": "Person",
  "name": "User Person",
  "organization": {
    "id": 1,
    "orgId": 1,
    "organizationBannerUrl": null,
    "organizationDescription": "UBC Okanagan",
    "organizationLogoUrl": null,
    "organizationName": "UBCO",
    "organizationRole": "member",
  },
  "photoURL": "https://example.com",
  "readChangeLog": false,
  "sid": null,
  "userRole": "user",
}
`;
